src/python/pants/backend/python/macros/poetry_project.py

```diff
@@ -0,0 +1,196 @@
+
+    git_lookup = attributes.get("git")
+    if git_lookup is not None:
+        # Making revision optional and supporting a branch are not
+        # strictly in agreement with PEP440 (does not comment on
+        # specifying a branch). However, pip seems to accept it fine
+        # (see pypa/pip issue 7650)
+        # so I can't see a reason not to support it here...?
+
+        rev_lookup = produce_match("#", attributes.get("rev"))
+        branch_lookup = produce_match("@", attributes.get("branch"))
+        tag_lookup = produce_match("@", attributes.get("tag"))
+
+        return f"{proj_name} @ git+{git_lookup}{tag_lookup}{branch_lookup}{rev_lookup}"
+
+    version_lookup = attributes.get("version")
+    path_lookup = attributes.get("path")
+    if path_lookup is not None:
+        return f"{proj_name} @ file://{path_lookup}"
+    url_lookup = attributes.get("url")
+    if url_lookup is not None:
+        return f"{proj_name} @ {url_lookup}"
+    if version_lookup is not None:
+        markers_lookup = produce_match(";", attributes.get("markers"))
+        python_lookup = parse_python_constraint(attributes.get("python"))
+        version_parsed = handle_str_attr(proj_name, version_lookup)
+        return (
+            f"{version_parsed}"
+            f"{markers_lookup}"
+            f"{' and ' if python_lookup and markers_lookup else (';' if python_lookup else '')}"
+            f"{python_lookup}"
+        )
+    else:
+        # This case should never happen if input is formed correctly; this
+        # ultimately will cause the Requirement.parse call to fail.
+        # TODO: add appropriate error handling
+        return ""
+
+
+def parse_single_dependency(
+    proj_name: str, attributes: str | dict[str, Any] | list[dict[str, Any]]
+) -> tuple[Requirement, ...]:
+    if isinstance(attributes, str):
+        return (Requirement.parse(handle_str_attr(proj_name, attributes)),)
+    elif isinstance(attributes, dict):
+        return (Requirement.parse(handle_dict_attr(proj_name, attributes)),)
+    elif isinstance(attributes, list):
+        return tuple([Requirement.parse(handle_dict_attr(proj_name, attr)) for attr in attributes])
+    else:
+        # Similarly, this is currently added just for mypy checks. Error
+        # handling needs to be written to address this properly.
+        return (Requirement.parse(""),)
+
+
+def parse_pyproject_toml(toml_contents: str) -> set[Requirement]:
+    parsed = toml.loads(toml_contents)
+    poetry_vals = parsed["tool"]["poetry"]
+    try:
+        dependencies = poetry_vals["dependencies"]
+        dev_dependencies = poetry_vals["dev-dependencies"]
+    except KeyError:
+        print(
+            "Missing required `poetry.tools.dependencies` and/or `poetry.tools.dev-dependencies`"
+            "fields...perhaps you specified the wrong file?"
+            "These can be empty if, for example, you have no dev-dependencies."
+        )
+        return set()
```

Comment:
Here, I think the more interesting case is no requirements being defined, rather than the section headers not existing. Instead of checking for a `KeyError`, you can use `poetry_vals.get("dependencies", {})` etc, combine both like you do with `raw_reqs` now, then do `if not raw_reqs`. 

I think then log a warning, given that we don't error on an empty `requirements.txt`:

> No requirements defined in `poetry.tools.dependencies` and `poetry.tools.dev-dependencies` in {file_path}, which is loaded by Pants from a poetry_requirements macro. Did you mean to add some requirements?